ql/src/java/org/apache/hadoop/hive/ql/DriverContext.java

```diff
@@ -247,4 +275,107 @@ public String getQueryErrorMessage() {
+   * @param tree the root ASTNode of the query
+   */
+  public void setQueryType(BaseSemanticAnalyzer sem, ASTNode tree) {
+    List<Task<? extends Serializable>> rootTasks = sem.getAllRootTasks();
+    QueryProperties props = sem.getQueryProperties();
+    queryType = isAnalyze(props) ? "STATS" : (isDML(props) ? "DML" : null);
+    // common confusion whether CTAS is DML or DDL, let's pick DDL here
+    if (props != null && props.isCTAS()) {
+      queryType = "DDL";
+    }
+    if (queryType != null) {
+      return; //already figured out
+    }
+    queryType = rootTasks.stream().findFirst().map(t -> {
+      String type = t.getType().toString();
+      // a MAPRED stage could mean an INSERT query also
+      if (props != null && type.equalsIgnoreCase("MAPRED")) {
+        return props.isInsert ? "DML" : "QUERY";
+      }
+      return type;
+    }).orElseGet(() -> {
+      // in case of a semantic exception (e.g. a table not found or something else)
+      // the root AST Node can still imply if this is a query, try to fall back to that
+      // instead of "UNKNOWN"
+      if (tree.getText().equalsIgnoreCase("TOK_QUERY")){
+        return "QUERY";
+      }
+      // if there is a fetch task, this is a query
+      if (sem.getFetchTask() != null) {
+        return "QUERY";
+      }
+      // from this point, best efforts come
+      // whether it's a DDL?
+      if (DDLSemanticAnalyzerFactory.handles(tree)) {
+        return "DDL";
+      }
+      // whether it's DML? UPDATE/DELETE/MERGE queries are handled with an instance of RewriteSemanticAnalyzer
+      if (RewriteSemanticAnalyzer.class.isAssignableFrom(sem.getClass())) {
+        return "DML";
+      }
+      return "UNKNOWN";
+    });
+
+    // in case of DDL queries, the DDL type can be figured out from the tree root token, e.g.:
+    // CREATETABLE, ALTERTABLE_ADDPARTS, SHOWDATABASES, SHOWTABLES, etc.
+    this.ddlType = "DDL".equalsIgnoreCase(queryType) ? tree.getText().substring("TOK_".length()) : "";
+  }
+
+  /**
+   * The return value of either isAnalyzeCommand() or isAnalyzeRewrite() is always true for analyze commands:
+   * isAnalyzeCommand=true for "compute statistics",
+   * isAnalyzeRewrite=true for "compute statistics for columns".
+   * @return
+   */
+  private boolean isAnalyze(QueryProperties queryProperties) {
+    return queryProperties != null && (queryProperties.isAnalyzeCommand() || queryProperties.isAnalyzeRewrite());
+  }
+
+  /**
+   * DML (Data Manipulation Language) operations are operations that modify data,
+   * queryProperties.isInsert seems to reflect this very well: true for INSERT [OVERWRITE], UPDATE, DELETE, MERGE
+   * @return
+   */
+  private boolean isDML(QueryProperties queryProperties) {
+    return queryProperties != null && queryProperties.isInsert;
+  }
+
+  public boolean isQueryHistoryExplainEnabled() {
+    return conf.getBoolVar(HiveConf.ConfVars.HIVE_QUERY_HISTORY_SERVICE_ENABLED) &&
```

Comment:
why do you need to evaluate it every time from conf + not even sure it should be here

Reply:
ack, this can be a field

Reply:
+ moved the convenience method to the HiveConf (which already has similar ones like isWebUiQueryInfoCacheEnabled)